src/watchman_client.js

```diff
@@ -0,0 +1,341 @@
+'use strict';
+
+var watchman = require('fb-watchman');
+
+/**
+ * Constants
+ */
+
+/**
+ * Singleton that provides a public API for a single connection to a watchman instance for 'sane'.
+ * It tries to abstract/remove as much of the boilerplate processing as necessary
+ * from WatchmanWatchers that use it. In particular, they have no idea whether
+ * we're using 'watch-project' or 'watch', what the 'project root' is when
+ * we internally use watch-project, whether a connection has been lost
+ * and reestablished, etc. Also switched to doing things with promises and known-name
+ * methods in WatchmanWatcher, so as much information as possible can be kept in 
+ * the WatchmanClient, ultimately making this the only object listening directly
+ * to watchman.Client, then forwarding appropriately (via the known-name methods) to 
+ * the relevant WatchmanWatcher(s).
+ *
+ * @class WatchmanClient
+ * @param String dir
+ * @param {Object} opts
+ * @public
+ */
+
+function WatchmanClient() {
```

Comment:
This PR should likely match structural design choices of other parts of the code-base. Specifically as you can see in https://github.com/amasad/sane/blob/master/src/fsevents_watcher.js#L39 it follows:

* public instance state is `this.value = value`
* private instance state is `this._value = value`
* public methods are `ClassName.prototype.method`
* private methods are `ClassName.prototype._method`

Apply the above rules to this PR will:

* ensure consistent style
* reduce function nesting.
* easy debuggability (closures can make it harder to inspect internal "hidden" state.

Reply:
Agree with the naming mechanism. One question: in existing WatchmanWatcher, init() should really be private, since (as far as I know) it shouldn't be called from elsewhere. Any idea why it isn't _init()?

Reply:
Changed to _init().

Reply:
Also, changed to using object member vars instead of hidden internal state/functions. Now using _ for internal state, functions.